### DISCUSSION TITLE
Automatic timezone conversion in Eloquent?

### DISCUSSION THREAD

**Discussion by redelschaap** [laravel/framework#53629] [Ideas]
I am developing an application that has an API where consumers can send requests containing dates. They would post these dates in ISO format, containing a timezone reference. Currently, we have to do a conversion ourselves to the application's timezone (typically UTC). The framework does not convert the timezone when storing dates into the database. To be honest, I would expect this kind of magic from Laravel.

Let's imagine a model with a date cast:

```php
class BlogPost extends Model {
    protected $casts = [
        'published_at' => 'datetime',
    ];
}
```

When inserting a model from an API request that accepts an ISO formatted datetime string, we need to store this as follows:

```php
BlogPost::create([
    // ...
    'published_at' => Carbon::parse($request['publishedAt'])->setTimezone(config('app.timezone'));
    // ...
]);
```

It would be very nice, and would feel much more like Laravel, when we can leave out the `setTimezone()` part.

Now, I know that this has already been discussed before and until now the message was that this behavior was intended. But with all the good magic Laravel does on other parts of the framework, it really feels weird that It lacks this kind of magic. At least for me, and given the number of questions about this, also for many other developers.

Would the framework be open for automatic timezone conversion? Perhaps in the next major release, since this will probably break some applications. Although it should only break things when developers currently adjust time or anticipate on this behavior in another way than setting the timezone themselves. I can also think of an option where this would be an opt-in or opt-out function, just like with the `preventLazyLoading` or `preventSilentlyDiscardingAttributes` features.

I am willing to write a PR, but only when it has a chance of getting accepted of course. Please let me know!

**Comment by shaedrich** [comment]
When you use form requests, you can do

```php
class MyFormRequest
{
    public function rules(): array
    {
        return [
            // …
        ];
    }

    public function passedValidation()
    {
        $this->merge([
            'publishedAt' => $this->date('publishedAt')->setTimezone(config('app.timezone')),
        ]);
    }
}
```

I know, it's not a solution, but it can be a workaround, especially if you use this in a base class.

**Reply by redelschaap** [reply]
Of course there are multiple ways to make this a bit easier, but nothing would be as easy as when it "just works" out of the box.

**Reply by shaedrich** [reply]
I guess, there are several road blocks:
* Carbon is not Laravel, even though, Laravel has a wrapper for it. Yet, Carbon does essentially not know about Laravel's configuration
* while FormRequests have a `prepareForValidation()` and `passedValidation()` method, there currently is no declarative way to sanitize/normalize/cast/whatever the data before validation like the `rules()` method allows us for the data's validation. So, to Laravel, your date is just a string in the request itself

This makes there places where automatic conversation _could_ take place and Laravel _can safely_ know from which timezone to which timezone to convert very limited.

Though, I do like your suggestion of making this an opt-in or opt-out function.

**Reply by redelschaap** [reply]
Personally, I don't think that Carbon needs to know about Laravel, but Laravel should handle Carbon objects (or DateTime objects for that matter) in a way that dates in other timezones are stored in a logical way.

In Eloquent related code, checks are already performed to check whether a value is a DateTime or Carbon object. I think it could be easily added to that code. That way, (form) request related code does not have to change. After all, there are other ways to get data containing dates in other timezones, for example when retrieving information from an external API.

I'm still wondering if this would be something that has a chance of getting accepted. Perhaps one of the maintainers can clarify this (maybe @driesvints)?

**Reply by driesvints** [reply]
I've talked about this many times before on older issues but Laravel does not and will never do automatic timezone conversions for you. You are responsible for storing time in the timezone you want in your database and preferable that should be UTC. If we were to adjust the behaviour then it would break tons of Laravel applications and people who expect the current behaviour will want it back 😅 
If we introduce another subset of features to still offer automatic timezone conversion then that would just lead to more confusion.

Rule of thumb: don't change your `app.timezone` setting and leave it alone.

**Reply by Sophist-UK** [reply]
@driesvints Dries

I think we all understand the need for backwards compatibility so anything like this would need to be fully backwards compatible. I note on further research that some databases support blueprint dateTimeTz and timestampTz fields, so database support is probably already covered. Hence my suggestion below to provide standard casts for this (that would have to be explicitly added to a model) and which would therefore be fully backwards compatible for existing apps. But I can see no reason why this cannot be added using a package.

**Reply by redelschaap** [reply]
@driesvints Well, I'm not really talking about cases where the `app.timezone` has changed. I agree that the timezone should stay on UTC. When it comes to storing dates in the database, the current behavior is actually confusing to me as a developer.

Let me try to demonstrate for a final attempt. Let's say I retrieved a datetime value in another timezone and I want to store this date on a model:
```php
$date = Carbon::parse('2024-12-04T12:34:56+02:00');

$blogPost = BlogPost::create([
  'title' => 'Lorem ipsum dolor sit amet',
  'published_at' => $date
]);
```



Dumping some variables will give me the following result:

```
BlogPost {
  id: 1
  title: "Lorem ipsum dolor sit amet"
  published_at: "2024-12-04 12:34:56"
  created_at: "2024-12-04 20:26:35"
  updated_at: "2024-12-04 20:26:35"
}

$date->toIso8601String() => "2024-12-04T12:34:56+02:00"
$date->setTimezone('UTC')->toIso8601String() => "2024-12-04T10:34:56+00:00"
$blogPost->published_at->toIso8601String() => "2024-12-04T12:34:56+00:00"
```

Storing the datetime value in `published_at` like this will loose the timezone information. Since the app timezone is set to UTC, I would expect datetimes to be converted to UTC. However, the next time I retrieve the model, the timezone is just discarded and the time is 2 hours off.

---

I understand that a change like this would break tons of applications. That's why I suggested an opt-in like the `preventSilentlyDiscardingAttributes` feature. But let's be honest, with all the magic Laravel brings, we can at least expect timezone conversion when storing datetime values from different timezones, right?

**Reply by driesvints** [reply]
@redelschaap You need to do:

```
$blogPost = BlogPost::create([
  'title' => 'Lorem ipsum dolor sit amet',
  'published_at' => $date->toUtc(),
]);
```

And if you want to remember the timezone it originally was you need to do:

```
$blogPost = BlogPost::create([
  'title' => 'Lorem ipsum dolor sit amet',
  'published_at' => $date->toUtc(),
  'published_at_timezone' => $date->timezone,
]);
```

Then re-apply it on retrieval if you want it to be in the same timezone again.

> Since the app timezone is set to UTC, I would expect datetimes to be converted to UTC.

That's simply not how things work. It's intended not to. 

> But let's be honest, with all the magic Laravel brings, we can at least expect timezone conversion when storing datetime values from different timezones, right?

No, we cannot 😅 this is just a wrong assumption sorry. There are tons of apps out there who simply don't take the timezone of a datetime object into account and just work with the date and time of the values without expecting it to convert at all. 

`app.timezone` isn't and never was intended to convert time stored on Eloquent models. In fact it has nothing to do with Eloquent at all. The docs nowhere claim it either. This setting is purely for new datetime objects being created in your app and for things like queues, the task scheduler and console commands.

**You are responsible yourself for making sure you store datetime objects in the timezone you expect.**

**Reply by Sophist-UK** [reply]
Whilst I agree with @driesvints Dries here about responsibility for handling timezones correctly, the entire purpose of frameworks like Laravel and their add-on package ecosystems is that they provide standard functionality to avoid web developers from having to reinvent the same wheel time and time again.

Whilst it is absolutely the responsibility of each web developer to handle timezones according to their own needs - and sites can have very different needs preventing any sort of default handling - that doesn't mean that Laravel (and the wider package ecosystem) shouldn't provide some standard functionality and documentation to make it much easier for web developers to do this right (and see my comments below).

I don't think that it matters that much whether this functionality and documentation is delivered using a package or by Laravel itself - and Taylor's policy is for this type of thing to be delivered using packages unless it becomes such common usage that it should be rolled into Laravel itself.

So, if there is consensus that such a package is needed, let's stop lobbying the Laravel team to include in Laravel and instead let's find someone willing to write and maintain a suitable package that provides the casts and any other common functions together  with documentation to save people reinventing the wheel here.

As already offered below, if I ever get the time to write my open source Laravel project then I will need this functionality myself, and if a suitable package doesn't exist at that point I will write something and spin it off as a separate and independent open source package. But if anyone else wants to create such a package in the meantime, they should go for it.

**Reply by redelschaap** [reply]
Whilst I agree that it is the developer's responsibility to make sure dates are stored in the correct timezone, I still think the framework could offer a solution to make this easier.

I think this is very similar to the missing attributes. It is the developer's responsibility to not use any attributes that don't exist. However, the framework offers a way to make this easier to prevent.

But I'll accept that the maintainer have a different opinion about this. If I have the time, I am also open to make this as a package.

**Comment by Sophist-UK** [comment]
In a Joomla plugin I used to use a decade or so ago, timezone handling was one of the biggest cause of issues and the biggest headache for developers to get right.

In my opinion:

* If Carbon doesn't already have date-time conversions between time zones (handling daylight savings times based on either the date & time if the date is provided or a date provided separately or today's date if not date provided), then this should be added.
* Laravel should have functionality as standard for casting date-times to Zulu to store in a database and (obviously) in reverse when retrieving from the database plus built-in functionality to hold and store a time-zone alongside a database date-time field plus automatic conversion to/from this stored timezone or to a user's profile / session time-zone where a timezone is not stored alongside.

I haven't got the time to check whether these functions are already provided in Carbon or Laravel.

**Reply by Sophist-UK** [reply]
From a Laravel perspective this can almost certainly be provided by a package - and at some point I might even get around to writing it.

**Reply by redelschaap** [reply]
Actually, I tink dates should just be stored in the timezone configured in `app.timezone`, which is set to `UTC` for most applications. This keeps the values simple (in one format) and this does not require any other level of database support.

My point is that the framework could easily do the `->setTimezone('UTC')` calls that we now have to do manually. Carbon already has timezone conversion (handling daylight savings times) out of the box, so that part is covered.

**Reply by Sophist-UK** [reply]
See @driesvints Dries answer above as to why app.timezone is intended for a different usage.

> My point is that the framework could easily do the `->setTimezone('UTC')` calls that we now have to do manually.

This would be completely backwardly incompatible with all existing databases that haven't done this manually already. So it is never going to happen.

We just need to make timezone conversions much easier - by which I mean to find a way to make single explicit declaration in the model that:

1. If the timezone is not explicitly stored in the database alongside the datetimestamp, define the timezone that is used for the data stored in the database.

2. Define the Model's field's timezone to be used for automatic conversion back and forth either using a web site default (which could be `app.timezone()` or similar), or e.g. a session variable that holds the user specific timezone.

3. Undertaking the automatic conversion between database and model. This would normally be done in Laravel using casts. 

4. Store the local timezone that the model now uses as a separate attribute in the model (separately from the attribute for any timezone stored in the database). This might best be done by sub-classing Carbon classes to hold this information and to provide methods to convert between timezones.

5. Handles all the Blueprint date-time field types extremely simply, equally well and in a consistent manner:
   * dateTimeTz
   * dateTime
   * date
   * timeTz
   * time
   * timestampTz
   * timestamp
   * timestampsTz
   * timestamps

6. Handles all the variant use cases including backwards compatibility for data already stored (i.e. stored in `app.timezone` values), difficulties of meeting times used by people in different timezone possibly one with and one without DST (since DST changes in spring & autumn occur on different dates in different countries), handles DST or not as requested including both current algorithmic knowledge of DST change dates and a historical database of DST change dates where they were historically different) etc. etc. etc. (IMO THIS IS NON TRIVIAL!!!!)

7. Existing code would need to continue to work as it does now - and if a developer wants to use this comprehensive timezone functionality then they will need to explicitly code for it. The objective is to make this explicit usage as simple and automated as possible.

**Comment by KanekiYuto** [comment]
Why not store the timestamps and cast them when you retrieve them

**Comment by qf-cavila** [comment]
I guess what is being discussed is that the application could store data in the database using UTC, while allowing the application itself to operate in a different timezone.

There should be an automatic conversion layer that handles the translation between the storage timezone and the application timezone. This conversion should occur both when retrieving and persisting data.

The layer should check the timezone of the Carbon instance and, if it differs from the expected timezone, it should automatically perform the necessary conversion — converting to UTC when saving, and converting to the application timezone when reading from the database.

That would solve the problem pointed.

**Comment by Sandalanka** [comment]
I see exactly what you mean 👍 — this is a common pain point in Laravel when dealing with datetimes coming in from clients in arbitrary timezones (usually ISO8601 strings with offsets).

Currently, Eloquent’s $casts = ['published_at' => 'datetime'] does parse the incoming string into a Carbon instance correctly, but it doesn’t normalize the timezone before saving. That means whatever offset was in the incoming string is persisted as-is in the database, which makes cross-timezone consistency tricky.

Why Laravel behaves this way now

Laravel’s philosophy here has been to not make assumptions about how you want to persist datetimes.

Some applications do want to store datetimes in local timezones (e.g. for auditing, or where the user’s local time matters), so automatic conversion to app.timezone could silently break apps.

That’s why the official guidance has been: if you want UTC (or config('app.timezone')), you normalize explicitly before saving.

How this could be improved

I think your suggestion makes sense and could be accepted — if it’s done in a way that doesn’t silently break existing applications. A few possible approaches:

Global opt-in config

// config/app.php
'normalize_datetime_on_save' => true,


If enabled, then all datetime casts normalize to app.timezone before persisting.

Cast modifier
Something like:

protected $casts = [
    'published_at' => 'datetime:normalized',
];


So developers opt-in on a per-attribute basis.

Custom cast (today’s workaround)
Until Laravel natively supports it, you can implement your own:

use Illuminate\Contracts\Database\Eloquent\CastsAttributes;
use Carbon\Carbon;

class NormalizedDatetime implements CastsAttributes
{
    public function get($model, string $key, $value, array $attributes)
    {
        return $value ? Carbon::parse($value)->timezone(config('app.timezone')) : null;
    }

    public function set($model, string $key, $value, array $attributes)
    {
        return $value ? Carbon::parse($value)->setTimezone(config('app.timezone')) : null;
    }
}


Then in your model:

protected $casts = [
    'published_at' => NormalizedDatetime::class,
];

Would a PR be accepted?

Based on how Taylor & the Laravel core team have historically handled this, here’s my take:

Silent global change (i.e. always normalizing automatically) → 🚫 almost certainly won’t be accepted because it would break apps that expect current behavior.

Config flag or opt-in cast type → ✅ much more likely to be accepted, because it doesn’t break existing apps and gives developers the “Laravel magic” if they want it.

So if you want to open a PR, I’d suggest proposing an opt-in feature (either per-cast or via config), not a global change. That way it aligns with other Laravel “strict mode” features like preventLazyLoading().
